tools/codegen/gen.py

```diff
@@ -327,7 +337,118 @@ def func(f: NativeFunction) -> Optional[str]:
         else:
             assert_never(target)
 
-    return func
+    def structured_func(g: NativeFunctionGroup) -> List[str]:
+        assert g.out is not None
```

Comment:
Would it make sense at some point to impose that all 3 variants of a NativeFunctionGroup exist (or if not, are there explicit reasons to not want an op to have all 3 variants)? Probably not at this point (especially if that's not the case for a lot of ops), but maybe once more/most ops have migrated to structured kernels?

Reply:
I think it makes sense to always require an `_out` variant; we don't require this right now and so there are a bunch of functions that are missing `_out` variants when they should have them.

Inplace variant doesn't make sense to enforce though; many functions can't do inplace at all because their output size doesn't match any of their input sizes.

Reply:
What about requiring at least `functional`? That seems like the canonical member (at least, I can't think of any cases that lack one). 

If that axiom holds, there's the added benefit of being able to conceptually think of a NFG as a canonical definition + two optional variants... also means no need to do dynamic checking for odd cases like `NativeFunctionGroup(None, None, None)`.

This is kind of monday morning quarterbacking on NFG, sorry about that.

Reply:
In the latest version of the patch these are all required now.